File: ios/Runner/SdkSingleton.swift
//
//  InitSDK.swift
//  Runner
//
//  Created by Lukas Körbler on 11.12.23.
//

import Foundation
import BluecodeSdk

struct BlueCodeProps {
    public var sdkHost: String;
    public var appScheme: String;
    public var deepLink: String;
    public var isProduction: Bool;
    
    init(from dictionary: [String: Any]) {
        self.sdkHost = dictionary["sdkHost"] as? String ?? "SAMPLEAPP"
        self.appScheme = dictionary["appScheme"] as? String ?? "sampleapp"
        self.deepLink = dictionary["deepLink"] as? String ?? ""
        self.isProduction = dictionary["isProduction"] as? Bool ?? false
    }
     
    func toConfig() -> BCUiConfig{
        var config = BCUiConfig();
        config.sdkHost = self.sdkHost;
        config.appScheme = self.appScheme;
        config.sdkEnvironment = self.isProduction ? .production : .sandbox;
        config.supportedLanguages = [.de, .en]
        config.domain = .BCE
        config.enableWelcomePage = true
        config.sendNotificationOnPaymentSuccess = true
        config.enableLegalPage = true
        return config;
    }
}

@objc class SdkSingleton: NSObject {
    @objc static let shared = SdkSingleton()
    private var isBluecodeInitialized = false
    private var sdkEnviroment: BCUiEnvironment? = nil
    private var isSdkVisible: Bool = false

    @objc static func initBluecodeSDK(blueCodeProps: [String: Any]) {
        SdkSingleton.shared.ensureInitBluecodeSDK(blueCodeProps: blueCodeProps, customLayoutProviderDelegate: nil)
    }
    
    @objc func ensureInitBluecodeSDK(blueCodeProps: [String: Any], customLayoutProviderDelegate: BluecodeSdk.BCUiCustomLayoutProvider?) {
        
        let props = BlueCodeProps.init(from: blueCodeProps);

        if(!isBluecodeInitialized){
            let config: BCUiConfig = props.toConfig();
            sdkEnviroment = config.sdkEnvironment;
            
            BCUiSdk.intialize(config: config);
            BCUiSdk.customLayoutProvider = customLayoutProviderDelegate;
            isBluecodeInitialized = true
        }
        
        BCUiSdk.customLayoutProvider = customLayoutProviderDelegate;
        
        if(!props.deepLink.isEmpty){
              BCUiSdk.urlScheme = props.deepLink;
        }
    }
    
    @objc static func isInitialized() -> Bool {
        return shared.isBluecodeInitialized
    }
    
    @objc static func setUrlScheme(deepLink: String) {
        BCUiSdk.urlScheme = deepLink;
    }
    
    @objc func getSdkEnviroment() -> BCUiEnvironment {
        return sdkEnviroment!;
    }
    
    @objc func checkIfVisible() -> Bool {
        return isSdkVisible
    }
  
    @objc func BluecodeVisible() {
        isSdkVisible = true
    }

    @objc func BluecodeNotVisible() {
        isSdkVisible = false
    }
    
    @objc func resetSdk() {
        BCUiSdk.reset();
    }
}

File: android/app/src/main/kotlin/com/codetribe/alp_pay_flutter/SdkSingleton.kt
package com.codetribe.alp_pay_flutter

import android.app.Application
import android.os.Bundle
import at.bluecode.sdk.bluecodesdk.public.BCUiSdk
import at.bluecode.sdk.bluecodesdk.public.models.BCUiConfig
import at.bluecode.sdk.bluecodesdk.public.models.BCUiEnvironment
import at.bluecode.sdk.bluecodesdk.public.models.BCUiLanguage
import at.bluecode.sdk.token.BCDomain
import at.bluecode.sdk.bluecodesdk.public.provider.BCUiCustomLayoutProvider

class SdkSingleton private constructor(private val application: Application) {
    private var isBluecodeInitialized = false
    private var isSdkVisible = false

    companion object {
        @Volatile private var instance: SdkSingleton? = null

        fun getInstance(application: Application): SdkSingleton {
            return instance ?: synchronized(this) {
                instance ?: SdkSingleton(application).also { instance = it }
            }
        }
        fun isInitialized(): Boolean {
            return instance?.isBluecodeInitialized ?: false
        }

        fun setUrlScheme(deepLink: String?) {
            BCUiSdk.setUrlScheme(deepLink);
        }

        fun checkIfVisible(): Boolean? {
            return instance?.isSdkVisible
        }

        fun BluecodeVisible() {
            instance?.isSdkVisible = true
        }

        fun BluecodeNotVisible() {
            instance?.isSdkVisible = false
        }
        fun resetSdk() {
            BCUiSdk.reset();
        }
    }

    fun initBluecodeSDK(blueCodeConfig: Bundle?) {
        ensureInitBluecodeSDK(blueCodeConfig, null)
    }

    fun ensureInitBluecodeSDK(
        blueCodeConfig: Bundle?,
        customLayoutProviderDelegate: BCUiCustomLayoutProvider?
    ) {
        val props = BluecodeConfig(blueCodeConfig)
        val config = props.toConfig()

        if (!isBluecodeInitialized) {
            BCUiSdk.initialize(application, config)
            isBluecodeInitialized = true
        }

        if (props.deepLink.isNotEmpty()) {
            BCUiSdk.setUrlScheme(props.deepLink)
        }

        customLayoutProviderDelegate?.let {
            BCUiSdk.setCustomLayoutProvider(it)
        }
    }



    class BluecodeConfig(bluecodeConfig: Bundle?) {
        var sdkHost = bluecodeConfig?.getString("sdkHost") ?: "SAMPLEAPP";
        var appScheme = bluecodeConfig?.getString("appScheme") ?: "sampleapp";
        var deepLink = bluecodeConfig?.getString("deepLink") ?: "";
        var isProduction= bluecodeConfig?.getBoolean("isProduction") ?: false;

        fun toConfig(): BCUiConfig{
            val config= BCUiConfig(
                sdkHost = this.sdkHost,
                appScheme = this.appScheme,
                sdkEnvironment = if(this.isProduction) BCUiEnvironment.PRODUCTION else BCUiEnvironment.SANDBOX,
                supportedLanguages = listOf(BCUiLanguage.EN, BCUiLanguage.EN),
                domain = BCDomain.BCE,
                enableWelcomePage = true,
                enableLegalPage = true,
                sendNotificationOnPaymentSuccess = true,
                appSupportsDarkMode = false,
            )
            return config;

        }
    }
}
